### PR TITLE
IEEE-12 Incident List API

### LINKED ISSUE #45: GET: /api/hardware/incidents/
List view of all hardware incidents, for use by the tech team admin panel.

- All fields
    - Also:
        - order_item
        - order__team
        - hardware__id
    - Filters:
        - By order_item__hardware__id
        - By order_item__order__team__id

### PR DESCRIPTION
**Pull request by Leo6Leo**
## Overview

- Resolves #45 
- List view of all hardware incidents, for use by the tech team admin panel.

All fields

Also:

order_item

order__team

hardware__id

Filters:

By order_item__hardware__id

By order_item__order__team__id

## Unit Tests Created

- Test for IncidentListView

### CODE REVIEW

File: hackathon_site/hardware/api_urls.py
Comment (grahamhoyes) [inline hackathon_site/hardware/api_urls.py]: Minor, but the `name` param here is typically singular form, so `name="incident-list"`

File: hackathon_site/hardware/serializers.py
Comment (grahamhoyes) [inline hackathon_site/hardware/serializers.py]: Minor, `IncidentSerializer` (singular form to match the rest)

File: hackathon_site/hardware/models.py
Comment (grahamhoyes) [inline hackathon_site/hardware/models.py]: As we talked about today, this isn't ideal. We can accomplish this in the serializer instead,  details below!

File: hackathon_site/hardware/serializers.py
Comment (grahamhoyes) [inline hackathon_site/hardware/serializers.py]: As we talked about:

```python
class OrderItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = OrderItem
        fields = ("id", "hardware", "order", "part_returned_health")


class IncidentsSerializer(serializers.ModelSerializer):
    order_item = OrderItemSerializer()
    team_id = serializers.SerializerMethodField()

    class Meta:
        model = Incident
        fields = (
            "id",
            "state",
            "time_occurred",
            "description",
            "order_item",
            "team_id",
            "created_at",
            "updated_at",
        )

    @staticmethod
    def get_team_id(obj: Incident):
        return obj.order_item.order.team.id
```

Comment (grahamhoyes) [reply hackathon_site/hardware/serializers.py]: Eventually, we'll need a custom field type for `team_id` that lets us use this serializer for POST requests as well, but that's a problem for another day.

File: hackathon_site/hardware/views.py
Comment (grahamhoyes) [inline hackathon_site/hardware/views.py]: Minor, `IncidentListView` (singular again)

File: hackathon_site/hardware/views.py
Comment (grahamhoyes) [inline hackathon_site/hardware/views.py]: For efficiency:

```python
    queryset = Incident.objects.all().select_related(
        "order_item", "order_item__order__team"
    )
```

File: hackathon_site/hardware/api_filters.py
Comment (grahamhoyes) [inline hackathon_site/hardware/api_filters.py]: `IncidentFilter` (singular)

File: hackathon_site/hardware/api_filters.py
Comment (grahamhoyes) [inline hackathon_site/hardware/api_filters.py]: These work great! But just to keep the field names and labels cleaner, can you do:

```python
    hardware_id = filters.NumberFilter(field_name="order_item__hardware__id")
    team_id = filters.NumberFilter(field_name="order_item__order__team__id")
```

Still works the exact same, just makes the query parameter cleaner and builds a label automatically

File: hackathon_site/hardware/serializers.py
Comment (grahamhoyes) [inline hackathon_site/hardware/serializers.py]: Will need a test for this serializer, in `hardware/tests.py`. Since there's nothing special going on in the serializer, you should be good with just one test method in the class where you hard-code the expected JSON response. Since time fields are annoying to fix in tests, take a look at [this](https://github.com/ieeeuoft/hackathon-template/blob/develop/hackathon_site/hardware/tests.py#L204) order serializer test for how to do it.

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py:385]: Cool! We probably need to fix a bunch of other tests to do that

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py]: Also needs test cases for your two filters:
- One that filters by hardware ID. You already have two hardware instances, you'll just need to create a second order item and incident.
- One that filters by team ID. You'll have to create a second team and order to do that, which you can tie to the second order item you'll need for testing the hardware filter.

**Review by grahamhoyes — COMMENTED** [review]
Fantastic start!

File: hackathon_site/hardware/serializers.py
Comment (Leo6Leo) [reply hackathon_site/hardware/serializers.py]: How does "get_team_id" know the value should go to the "team_id" field? Will Django automatically recognize the "team_id" in the "get_team_id" method? @grahamhoyes

Comment (grahamhoyes) [reply hackathon_site/hardware/serializers.py]: It's the other way around - by default, `SerializerMethodField` will look for a method called `get_<field name>`: https://www.django-rest-framework.org/api-guide/fields/#serializermethodfield

Comment (Leo6Leo) [reply hackathon_site/hardware/serializers.py]: got it! It makes a lot of sense rn. thanks Graham!

File: hackathon_site/hardware/views.py
Comment (grahamhoyes) [inline hackathon_site/hardware/views.py:60]: `SearchFilter` is unused here, since you also have to specify what fields can be searched on.

What do you think about these fields?

```python
    search_fields = (
        "state",
        "order_item__order__team__team_code",
        "order_item__hardware__name",
        "order_item__hardware__manufacturer"
    )
```

With those, update the queryset to be:

```python
    queryset = Incident.objects.all().select_related(
        "order_item", "order_item__order__team", "order_item__hardware"
    )
```

Then just one simple test case where you do

```python
url = self._build_filter_url(search="name")
```

and this will be good to go!

**Review by grahamhoyes — COMMENTED** [review]
Great work! Just need to make use of the `SearchField`

Sorry for taking so long to review this one 🙃

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py]: For the search filter, it's just `<url>?search=whatever`, and that will search all of the search fields you specify.

This test is actually a false-positive - since you have no filter arguments that look for `?name`, nothing happens to the results. But since there's only one `Incident` object created, this test doesn't capture that - and technically, none of the other ones do either.

Can you add a second `Incident` object to this test class? Your hardware filter test will be good to go then, and if you attach the incident to a different team (aka new order and order item as well), your team filter test will be good as well.

**Comment by Leo6Leo** [comment]
image

These mailer tests seem failed. All my tests have passed. 

image
And all error msgs are the same except for those two with the exclamation mark.

This is the error msg of the one with the exclamation mark.
image

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py]: It looks like Django is throwing a warning because it expects datetimes to be timezone aware. Easy fix:

```python
from datetime import datetime
...
from django.conf import settings
...

self.incident = Incident.objects.create(
    state="Broken",
    description="Description",
    order_item=self.order_item,
    time_occurred=datetime(2022, 8, 8, tzinfo=settings.TZ_INFO),
)
```

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py]: Same as above

File: hackathon_site/hardware/test_api.py
Comment (grahamhoyes) [inline hackathon_site/hardware/test_api.py]: An errant print!

File: hackathon_site/hardware/views.py
Comment (grahamhoyes) [inline hackathon_site/hardware/views.py:47]: Once #308 is merged, up to you if you want to add permissions into the view and tests now or if you want to do that as a separate PR

**Comment by grahamhoyes** [comment]
The mail tests work now, I reran the github pipeline. We have a know issue where they fail around midnight Toronto time - we don't know why or exactly what the time range is, and we don't know if they'll also fail if we run them in real life around midnight!

File: hackathon_site/hardware/test_api.py
Comment (Leo6Leo) [reply hackathon_site/hardware/test_api.py]: I have encountered this bug before, and when I set the year from 2021 to 2022, it just disappeared on my end.

File: hackathon_site/hardware/views.py
Comment (Leo6Leo) [reply hackathon_site/hardware/views.py:47]: I will make a separate PR to add the permissions to these views and tests when IEEE-69 is merged.

**Review by grahamhoyes — APPROVED** [review]
Nice work! 

As long as you hit the "Update branch" button in github, this should stay approved and you can merge it in.

**Pull request merged**
